kimchi/src/circuits/polynomials/foreign_field_add/circuitgates.rs

```diff
@@ -0,0 +1,233 @@
+//~ |   1 | `left_input_mi`  (copy) | `result_mi` (copy)     | `resmin_mi` (copy)   |
+//~ |   2 | `left_input_hi`  (copy) | `result_hi` (copy)     | `resmin_hi` (copy)   |
+//~ |   3 | `right_input_lo` (copy) |  ...                   | `bound_lo`  (copy)   |
+//~ |   4 | `right_input_mi` (copy) |  ...                   | `bound_mi`  (copy)   |
+//~ |   5 | `right_input_hi` (copy) |  ...                   | `bound_hi`  (copy)   |
+//~ |   6 | `field_overflow`        |  ...                   |  -                   |
+//~ |   7 | `carry_lo`              |  ...                   | `bound_carry_lo`     |
+//~ |   8 | `carry_mi`              |  ...                   | `bound_carry_mi`     |
+//~ |   9 | `sign`                  |  ...                   |  -                   |
+//~ |  10 |                         |                        |                      |
+//~ |  11 |                         |                        |                      |
+//~ |  12 |                         |                        |                      |
+//~ |  13 |                         |                        |                      |
+//~ |  14 |                         |                        |                      |
+//~
+//~ Having a specific final row that checks the bound is useful as it checks the upper bound
+//~ without reusing the same constraints in the `ForeignFieldAdd` rows (which would require
+//~ some extra constraints to be added to the circuit).
+
+/// Implementation of the ForeignFieldAddition gate
+pub struct ForeignFieldAdd<F>(PhantomData<F>);
+
+impl<F> Argument<F> for ForeignFieldAdd<F>
+where
+    F: FftField,
+{
+    const ARGUMENT_TYPE: ArgumentType = ArgumentType::Gate(GateType::ForeignFieldAdd);
+    const CONSTRAINTS: u32 = 7;
+
+    fn constraint_checks<T: ExprOps<F>>(env: &ArgumentEnv<F, T>) -> Vec<T> {
+        let field_overflow = env.witness_curr(6);
+        let sign = env.witness_curr(9);
+
+        let mut checks = vec![
+            // Field overflow bit is 0 or s.
+            field_overflow.clone() * (field_overflow - sign.clone()),
+            // Sign flag is 1 or -1
+            (sign.clone() + T::one()) * (sign - T::one()),
+        ];
+
+        // Carry bits are -1, 0, or 1.
+        checks.append(&mut carry(env));
+
+        // Result values match
+        checks.append(&mut result(env));
+
+        checks
+    }
+}
+
+/// Implementation of the FFFin gate
+pub struct ForeignFieldFin<F>(PhantomData<F>);
+
+impl<F> Argument<F> for ForeignFieldFin<F>
+where
+    F: FftField,
+{
+    const ARGUMENT_TYPE: ArgumentType = ArgumentType::Gate(GateType::ForeignFieldFin);
+    const CONSTRAINTS: u32 = 5;
+
+    fn constraint_checks<T: ExprOps<F>>(env: &ArgumentEnv<F, T>) -> Vec<T> {
+        // Similar to performing a foreign field addition gate with these inputs
+        // left_input_lo -> resmin_lo  right_input_lo -> 0     result_lo -> bound_lo  carry_lo -> bound_carry_lo
```

Comment:
For consistency with other variable naming in this area of the code, rename `resmin -> min_result_*`

Reply:
done